### PR TITLE
feat: support TextUnit on most sp/em properties

### PR DESCRIPTION
**Pull request by sargunv**
textOffset isn't supported yet; gotta think more on how to handle that one as apparently expressions aren't allowed inside list literals

### CODE REVIEW

**Comment by sargunv** [comment]
* **#140** <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/140?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/140?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt [hunk 1]
```diff
@@ -504,17 +520,39 @@ public fun SymbolLayer(
       set(iconTranslate) { layer.setIconTranslate(it) }
       set(iconTranslateAnchor) { layer.setIconTranslateAnchor(it) }
 
+      set(textSizeSp to fontScale) { (textSizeSp, fontScale) ->
+        layer.setTextSize((textSizeSp * const(fontScale)).dp)
+      }
```
Comment (westnordost) [inline lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:525]: I'd use `Pair(textSizeSp, fontScale)` for clarity (the `to` metaphor doesn't really work here), or, even better (IMO), use the `listOf` API consistently here when more than one parameter is involved.

(I do wonder why the `set` method doesn't simply have a vararg parameter)

**Review by westnordost — COMMENTED** [review]
API-wise, this looks pretty cool-

File: lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt (hunk 1 above)
Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:525]: I'm gonna use remember instead, and pass a single value in here. 

Simplified reimplementation coming in a bit

File: lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt [hunk 2]
```diff
@@ -428,27 +430,27 @@ public fun SymbolLayer(
...
+  textLineHeight: Expression<TextUnitValue> = const(1.2f.em),
   textJustify: Expression<EnumValue<TextJustify>> = const(TextJustify.Center),
   textWritingMode: Expression<ListValue<EnumValue<TextWritingMode>>> = nil(),
   textKeepUpright: Expression<BooleanValue> = const(true),
   textRotate: Expression<FloatValue> = const(0f),
 
   // text anchoring
   textAnchor: Expression<EnumValue<SymbolAnchor>> = const(SymbolAnchor.Center),
-  textOffset: Expression<OffsetValue> = const(Offset.Zero),
+  textOffset: Expression<OffsetValue> = const(Offset.Zero), // TODO text unit type
```
Comment (westnordost) [inline lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:451]: open TODO. (Didn't I read something about offsetX, offsetY just then? Also, there is `TextOffsetValue` but it is not used here?)

Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:451]: It turns out MLN doesn't support expressions inside list literals, so we can't split it into two expressions.

I'm updating the `literal` functions and then I'll try to find a new approach for dealing with the TextUnitOffset. Not sure if it'll be possible (can we divide an sp offset by an expression to get em?)

**Comment by westnordost** [comment]
Oh, **important**:

I noticed that `textVariableAnchorOffset` also has offsets in EM. The type is basically 

`List<Pair<SymbolAnchor, TextUnitOffset>>`

File: lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt (hunk 2 above)
Comment (westnordost) [reply lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:451]: if all fails, an option would be to revert back to `Float` for "EM" fields ¯\_(ツ)_/¯

**Comment by sargunv** [comment]
Yeah, I might consider that a `Map<SymbolAnchor< Expression<TextOffsetValue>>` and render the map as a list internally

**Comment by sargunv** [comment]
https://github.com/sargunv/maplibre-compose/pull/142 adds support for TextUnit in textOffset

**Comment by sargunv** [comment]
### Merge activity

* **Dec 18, 3:20 PM PST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/sargunv/maplibre-compose/140).

File: lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt (hunk 2 above)
Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:451]: It worked, #142 has it

Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/dev/sargunv/maplibrecompose/compose/layer/SymbolLayer.kt:451]: We convert from TextUnit to sp or em by multiplying it with some scalar coming from a variable bound in the symbol layer composable. To convert a 2D text unit, we need to multiply both components (x and y) with the scalar. We can't do expressions inside the list of x and y, so we need some way to multiply an entire list of numbers by some scalar.

MapLibre expressions don't provide this directly, but linear interpolation with stops above/below our scalar is effectively the same thing as multiplication by a scalar. And interpolation works on lists of numbers.

So if you assume, say, the SP to/from EM conversion will always multiply by a value in range 0..1000, we can define those stops on a linear interpolation, input the scalar, and we've got runtime conversion between SP/EM for 2d text units

**Pull request merged**